src/test/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreamsTest.java

```diff
@@ -545,20 +543,71 @@ void testLatest() {
     }
 
     /**
-     * Verifies that {@link ShardIteratorType#TRIM_HORIZON} requests do not cache iterators or records.
+     * Verifies that mapping from {@link ShardIteratorType#TRIM_HORIZON} to sequence number is cached.
      */
     @Test
-    void testTrimHorizonNotCached() {
+    void testTrimHorizonCached() {
         AmazonDynamoDBStreams streams = mock(AmazonDynamoDBStreams.class);
 
         CachingAmazonDynamoDbStreams cachingStreams = mockTrimHorizonStream(streams);
 
+        assertGetRecords(cachingStreams, newTrimHorizonRequest(), 10, 0, 10);
+
+        // verify that no iterator or records were requested from underlying stream
+        assertCacheMisses(streams, 1, 1);
+    }
+
+    /**
+     * Verifies that mapping from {@link ShardIteratorType#TRIM_HORIZON} to sequence number is cached only for a limited
+     * time.
+     */
+    @Test
+    void testTrimHorizonNotCached() {
+        AmazonDynamoDBStreams streams = mock(AmazonDynamoDBStreams.class);
+
+        MockTicker ticker = new MockTicker();
+
+        CachingAmazonDynamoDbStreams cachingStreams = mockTrimHorizonStream(streams, ticker);
+
+        // advance beyond default cache timeout of 60 seconds
+        ticker.increment(60, TimeUnit.SECONDS);
+
         assertGetRecords(cachingStreams, newTrimHorizonRequest(), null, 0, 10);
 
         // verify that underlying stream was still accessed for each request
         assertCacheMisses(streams, 2, 2);
     }
 
+    /**
+     * Verifies that records cache is used for TRIM_HORIZON request even if mapping was not yet cached when iterator was
+     * retrieved.
+     */
+    @Test
+    void testTrimHorizonCachedAfterGetIterator() {
+        AmazonDynamoDBStreams streams = mock(AmazonDynamoDBStreams.class);
+
+        GetShardIteratorRequest iteratorRequest = mockTrimHorizonRequest(streams, streamArn, shardId);
+
+        // get exact overlapping records with and without limit
+        CachingAmazonDynamoDbStreams cachingStreams = new CachingAmazonDynamoDbStreams
+            .Builder(streams)
+            .withTicker(new MockTicker())
+            .build();
+
+        // first thread looks up iterator
+        final String iterator = cachingStreams.getShardIterator(iteratorRequest).getShardIterator();
+        // second thread looks up iterator: no mapping cached, so results in call to DynamoDB
+        final String iterator2 = cachingStreams.getShardIterator(iteratorRequest).getShardIterator();
+
+        // first thread looks up records: no mapping cached, so results in call to DynamoDB
+        assertGetRecords(cachingStreams, iterator, null, records);
+        // second thread looks up records: mapping should be cached now, so even though first thread obtained phsyical
+        // iterator, its mapped to cached records without call to DynamoDB (limit so it doesn't try to look beyond)
+        assertGetRecords(cachingStreams, iterator2, 10, records);
+
+        assertCacheMisses(streams, 2, 1);
+    }
```

Comment:
@msgroi FYI, added another test case that was missing.

Reply:
👍